Snakefile

```diff
@@ -0,0 +1,402 @@
+    shell:
+        """
+        ./bin/transform-biosample {input.biosample} \
+            --output {output.biosample}
+        """
+
+rule transform_genbank_data:
+    input:
+        biosample = "data/genbank/biosample.tsv",
+        ndjson = "data/genbank.ndjson"
+    output:
+        fasta = "data/genbank/sequences.fasta",
+        metadata = "data/genbank/metadata_transformed.tsv",
+        duplicate_biosample = "data/genbank/duplicate_biosample.txt"
+    shell:
+        """
+        ./bin/transform-genbank {input.ndjson} \
+            --biosample {input.biosample} \
+            --duplicate-biosample {output.duplicate_biosample} \
+            --output-metadata {output.metadata} \
+            --output-fasta {output.fasta}
+        """
+
+rule transform_gisaid_data:
+    input:
+        ndjson = "data/gisaid.ndjson"
+    output:
+        fasta = "data/gisaid/sequences.fasta",
+        metadata = "data/gisaid/metadata_transformed.tsv",
+        flagged_annotations = temp("data/gisaid/flagged-annotations"),
+        additional_info = "data/gisaid/additional_info.tsv"
+    shell:
+        """
+        ./bin/transform-gisaid {input.ndjson} \
+            --output-metadata {output.metadata} \
+            --output-fasta {output.fasta}  \
+            --output-additional-info {output.additional_info} \
+            --output-unix-newline > {output.flagged_annotations};
+        """
+
+rule download_nextclade:
+    params:
+        dst_source=lambda w: config["s3_dst"] + '/nextclade.tsv.gz',
+        src_source=lambda w: config["s3_src"] + '/nextclade.tsv.gz',
+    output:
+        nextclade = "data/{database}/nextclade_old.tsv"
+    shell:
+        """
+        ./bin/download-from-s3 {params.dst_source} {output.nextclade} ||  \
+        ./bin/download-from-s3 {params.src_source} {output.nextclade}
+        """
+
+rule get_sequences_without_nextclade_annotations:
+    """Find sequences in FASTA which don't have clades assigned yet"""
+    input:
+        fasta = "data/{database}/sequences.fasta",
+        nextclade = "data/{database}/nextclade_old.tsv",
+    output:
+        fasta = "data/{database}/nextclade.sequences.fasta"
+    shell:
+        """
+        ./bin/filter-fasta \
+            --input_fasta={input.fasta} \
+            --input_tsv={input.nextclade} \
+            --output_fasta={output.fasta} \
+        """
+
+rule annotate_via_nextclade:
+    message:
+        """
+        If there are sequences without clades, then run nextclade to assign
+        clades and calculate other useful metrics.
+        """
+    input:
+        sequences = "data/{database}/nextclade.sequences.fasta",
+        nextclade_info = "data/{database}/nextclade_old.tsv"
+    params:
+        new_info = temp("data/{database}/nextclade_new.tsv"),
+        nextclade_input_dir = temp(directory("data/{database}/nextclade_inputs")),
+        nextclade_output_dir = temp(directory("data/{database}/nextclade"))
+    threads: 8
```

Comment:
Pre-merge change: We should pass this `threads` variable directly to Nextclade to prevent overcommitting resources when we run Snakemake rules in parallel (something we might not have been doing before with the bash scripts). We could do this by modifying the `run-nextclade` script to take a threads argument and then pass this with `run-nextclade --threads {threads}`.

Reply:
Good call - resolved in 8204f9b

Reply:
Hey @ivan-aksamentov - those changes were in response to this review comment so i'll inline responses here to keep things ordered.

> Will something run concurrently with Nextclade step?

Probably not at the moment, but I think @huddlej's comment was to future proof ourselves for the situation when a job does run concurrently.

> Where does the threads variable come from and what it will be by default?

Snakemake will choose the minumum of (i) the user-provided cores (via snakemake's `--cores` arg), (ii) system cores available and (iii) 8, which is hardcoded in the rule. I can easily change (iii) to 16. Relatedly, my understanding was that nextclade didn't really scale well beyond 8 cores, or was specific to nextalign?

P.S. I specifically modified the bash script in such a way that it's backwards compatible -- if you call it with 4 arguments as before, then the behavior is unchanged.

Reply:
Okay then. But we need to take care to not put unnecessary slowdowns/limits.


> Relatedly, my understanding was that nextclade didn't really scale well beyond 8 cores, or was specific to nextalign?

The scaling is not great currently, for both tools, but I don't think it should be additionally limited artificially. Especially considering that it is the slowest and most resource-intensive part of the pipeline. I think the hard limit of 8 or 16 threads should be removed. At least this is what I would do if I wanted things to go faster. I plan to invest some time on profiling and optimizing the C++ code, so that it will scale better hopefully soon.

Regarding the bash script, I think it's meant to go away anyways. It's really mostly a glue, with only 3 statements that do anything useful. So we should probably not add any more things into it. Taking care of backwards compat is also not necessary, I don't think anyone is using this script directly. But if keeping the current latest commit is the simplest for you, I'd just go with it.

Reply:
Sweet. I'll increase it to 16 cores (which is what the AWS job is provisioned with) before merge (but will first wait to see if the jobs succeed).

Agree with the statement that the bash script will go away eventually - my desire for this PR was to modify the status quo as little as possible.

Reply:
Resolved in 
416c56f